packages/rest-crud/src/crud-rest.controller.ts

```diff
@@ -0,0 +1,320 @@
+// Because the controller class is an internal type scoped to the body of the
+// factory function, we cannot use it to describe the return type. We must
+// explicitly provide the return type.
+//
+// To work around those issues, we use the following design:
+// - The interface `CrudRestController` describes controller methods (members)
+// - The type `CrudRestControllerCtor` describes the class constructor.
+// - `defineCrudRestController` returns `CrudRestControllerCtor` type.
+
+/**
+ * This interface describes prototype members of the controller class
+ * returned by `defineCrudRestController`.
+ */
+export interface CrudRestController<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+> {
+  /**
+   * The backing repository used to access & modify model data.
+   */
+  readonly repository: EntityCrudRepository<T, IdType>;
+
+  /**
+   * Implementation of the endpoint `POST /`.
+   * @param data Model data
+   */
+  create(data: Omit<T, IdName>): Promise<T>;
+}
+
+/**
+ * Constructor of the controller class returned by `defineCrudRestController`.
+ */
+export interface CrudRestControllerCtor<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+> {
+  new (
+    repository: EntityCrudRepository<T, IdType, Relations>,
+  ): CrudRestController<T, IdType, IdName, Relations>;
+}
+
+/**
+ * Options to configure different aspects of a CRUD REST Controller.
+ */
+export interface CrudRestControllerOptions {
+  /**
+   * The base path where to "mount" the controller.
+   */
+  basePath: string;
+}
+
+/**
+ * Create (define) a CRUD Controller class for the given model.
+ *
+ * Example usage:
+ *
+ * ```ts
+ * const CrudRestController = defineCrudRestController<
+ *   Product,
+ *   typeof Product.prototype.id,
+ *   'id'
+ * >(Product, {basePath: '/products'});
+ *
+ * class ProductController extends CrudRestController {
+ *   constructor() {
+ *    super(repo);
+ *   }
+ * }
+ *
+ * app.controller(ProductController);
+ * ```
+ *
+ * @param modelCtor A model class, e.g. `Product`.
+ * @param options Configuration options, e.g. `{basePath: '/products'}`.
+ */
+export function defineCrudRestController<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+>(
+  modelCtor: typeof Entity & {prototype: T & {[key in IdName]: IdType}},
+  options: CrudRestControllerOptions,
+): CrudRestControllerCtor<T, IdType, IdName, Relations> {
+  const modelName = modelCtor.name;
+  const idPathParam: ParameterObject = {
+    name: 'id',
+    in: 'path',
+    schema: getIdSchema(modelCtor),
+  };
+
+  @api({basePath: options.basePath, paths: {}})
+  class CrudRestControllerImpl
+    implements CrudRestController<T, IdType, IdName> {
+    constructor(
+      public readonly repository: EntityCrudRepository<T, IdType, Relations>,
+    ) {}
+
+    @post('/', {
+      ...response.model(200, `${modelName} instance created`, modelCtor),
```

Comment:
shall we change `200` to `201`? I remember many people expect to have `201` as the "successful created" code

Reply:
Eventually yes, but not yet. Responses with status code `201` are supposed to include `Location` header pointing to the newly created resource. We don't have a good solution allowing controller methods to specify a custom status code & response headers; plus it's non-trivial to calculate the correct `Location` - e.g. when the LB4 app is mounted on Express, or even worse when it's running behind a reverse proxy. See #2285, #2293 and #3133.

The scope of this task is to create REST API that's matching the API of controllers generated (scaffolded) via `lb4 controller`.

Adding support for 201 is out of scope, we already have a GH issue to track this feature - see https://github.com/strongloop/loopback-next/issues/788